holmes/plugins/toolsets/opensearch/opensearch_traces.py

```diff
@@ -0,0 +1,198 @@
+            if cached_response:
+                logging.debug("traces fields returned from cache")
+                return cached_response
+
+            body = {
+                "size": 1,
+                "_source": False,
+                "script_fields": {
+                    "all_fields": {
+                        "script": {
+                            "lang": "painless",
+                            "source": "Map fields = new HashMap(); List stack = new ArrayList(); stack.add(['prefix': '', 'obj': params['_source']]); while (!stack.isEmpty()) { Map item = stack.remove(stack.size() - 1); String prefix = item.get('prefix'); Map obj = item.get('obj'); for (entry in obj.entrySet()) { String fullPath = prefix.isEmpty() ? entry.getKey() : prefix + '.' + entry.getKey(); fields.put(fullPath, true); if (entry.getValue() instanceof Map) { stack.add(['prefix': fullPath, 'obj': entry.getValue()]); } } } return fields.keySet();",
+                        }
+                    }
+                },
+            }
+            headers = {"Content-Type": "application/json"}
+            headers.update(add_auth_header(self.toolset.config.opensearch_auth_header))
+            logs_response = requests.get(
+                url=self.toolset.config.opensearch_traces_url,
+                timeout=180,
+                verify=True,
+                data=json.dumps(body),
+                headers=headers,
+            )
+            logs_response.raise_for_status()
+            response = json.dumps(logs_response.json())
+            self._cache[TRACES_FIELDS_CACHE_KEY] = response
+            return response
+        except requests.Timeout:
+            logging.warn(
+                "Timeout while fetching opensearch traces fields", exc_info=True
+            )
+            return "Request timed out while fetching opensearch traces fields"
+        except RequestException as e:
+            logging.warn("Failed to fetch opensearch traces fields", exc_info=True)
+            return f"Network error while opensearch traces fields: {str(e)}"
+        except Exception as e:
+            logging.warn("Failed to process opensearch traces fields", exc_info=True)
+            return f"Unexpected error: {str(e)}"
+
+    def get_parameterized_one_liner(self, params) -> str:
+        return "list traces documents fields"
+
+
+class TracesSearchQuery(BaseOpenSearchTracesTool):
+    def __init__(self, toolset: "OpenSearchTracesToolset"):
+        super().__init__(
+            name="traces_in_range_search",
+            description="Get traces in a specified time range for an opensearch query",
+            parameters={
+                "query": ToolParameter(
+                    description="An OpenSearch search query. It should be a stringified json, matching opensearch search syntax. "
+                    "The query must contain a 'range' on the startTimeMillis field. From a given startTimeMillis, until a given startTimeMillis. This is time in milliseconds",
```

Comment:
Does this work well? Should holmes be given a tool to transform dates to millisecs?

Reply:
it works well on alerts, because I added those fields to the context
I'm calculating it when preparing the investigation context